### PR TITLE
x86-64: Use rep movs instructions if array copy size greater than 32 bytes

### PR DESCRIPTION
**Pull request by a7ehuo**
This change applies to byte and char array System.arraycopy sequence. 

It avoids using vector instructions for copy size between 32 bytes to 64 bytes due to an observed regression.

### CODE REVIEW

**Comment by a7ehuo** [comment]
@0xdaryl May I ask you to review this change? Thank you!

@vijaysun-omr @ymanton @hzongaro

File: compiler/x/codegen/OMRTreeEvaluator.cpp [hunk 1]
```diff
@@ -1891,7 +1891,11 @@ static void arrayCopy16BitPrimitiveInlineSmallSizeWithoutREPMOVSImplRoot16(TR::N
    // ---------------------------------
    generateLabelInstruction(TR::InstOpCode::label, node, copy18ORMoreBytesLabel, cg);
    generateRegImmInstruction(TR::InstOpCode::CMPRegImm4(), node, sizeReg, 32, cg);
-   generateLabelInstruction(TR::InstOpCode::JA4, node, copy34ORMoreBytesLabel, cg);
+
+   if (repMovsThresholdBytes == 32)
+      generateLabelInstruction(TR::InstOpCode::JA4, node, repMovsLabel, cg);
+   else
+      generateLabelInstruction(TR::InstOpCode::JA4, node, copy34ORMoreBytesLabel, cg);
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRTreeEvaluator.cpp:1898]: To reduce code footprint, please rewrite to something like:
```
TR::LabelSymbol *copyLabel = (repMovsThresholdBytes == 32) ? repMovsLabel : copy34ORMoreBytesLabel;
generateLabelInstruction(TR::InstOpCode::JA4, node, copyLabel, cg);
```

File: compiler/x/codegen/OMRTreeEvaluator.cpp [hunk 2]
```diff
@@ -1776,20 +1776,21 @@ static void arrayCopy16BitPrimitiveInlineSmallSizeWithoutREPMOVSImplRoot16(TR::N
                                                                            TR::Register* tmpXmmYmmReg1,
                                                                            TR::Register* tmpXmmYmmReg2,
                                                                            TR::CodeGenerator* cg,
+                                                                           int repMovsThresholdBytes,
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRTreeEvaluator.cpp:1779]: int32_t

File: compiler/x/codegen/OMRTreeEvaluator.cpp [hunk 3]
```diff
@@ -1900,27 +1904,74 @@ static void arrayCopy16BitPrimitiveInlineSmallSizeWithoutREPMOVSImplRoot16(TR::N
...
+   generateMemRegInstruction(TR::InstOpCode::VMOVDQUMemYmm, node, generateX86MemoryReference(dstReg, 0, cg), tmpXmmYmmReg2, cg);
    generateLabelInstruction(TR::InstOpCode::JMP4, node, mainEndLabel, cg);
+   }
+
+static void arrayCopy16BitPrimitiveInlineSmallSizeWithoutREPMOVSLargeSizeREPMOVSD(TR::Node* node,
+                                                                                  TR::Register* dstReg,
+                                                                                  TR::Register* srcReg,
+                                                                                  TR::Register* sizeReg,
+                                                                                  TR::CodeGenerator* cg,
+                                                                                  int repMovsThresholdBytes)
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRTreeEvaluator.cpp:1928]: int32_t

File: compiler/x/codegen/OMRTreeEvaluator.cpp [hunk 4]
```diff
@@ -2009,23 +2024,24 @@ static void arrayCopy8BitPrimitiveInlineSmallSizeWithoutREPMOVSImplRoot8(TR::Nod
                                                                          TR::Register *tmpXmmYmmReg1,
                                                                          TR::Register *tmpXmmYmmReg2,
                                                                          TR::CodeGenerator *cg,
+                                                                         int repMovsThresholdBytes,
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRTreeEvaluator.cpp:2027]: int32_t

File: compiler/x/codegen/OMRTreeEvaluator.cpp [hunk 5]
```diff
@@ -2135,7 +2151,11 @@ static void arrayCopy8BitPrimitiveInlineSmallSizeWithoutREPMOVSImplRoot8(TR::Nod
    // ---------------------------------
    generateLabelInstruction(TR::InstOpCode::label, node, copy17ORMoreBytesLabel, cg);
    generateRegImmInstruction(TR::InstOpCode::CMPRegImm4(), node, sizeReg, 32, cg);
-   generateLabelInstruction(TR::InstOpCode::JA4, node, copy33ORMoreBytesLabel, cg);
+
+   if (repMovsThresholdBytes == 32)
+      generateLabelInstruction(TR::InstOpCode::JA4, node, repMovsLabel, cg);
+   else
+      generateLabelInstruction(TR::InstOpCode::JA4, node, copy33ORMoreBytesLabel, cg);
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRTreeEvaluator.cpp:2158]: To reduce code footprint, please rewrite to something like:
```
TR::LabelSymbol *copyLabel = (repMovsThresholdBytes == 32) ? repMovsLabel : copy33ORMoreBytesLabel;
generateLabelInstruction(TR::InstOpCode::JA4, node, copyLabel, cg);
```

File: compiler/x/codegen/OMRTreeEvaluator.cpp [hunk 6]
```diff
@@ -2144,20 +2164,28 @@ static void arrayCopy8BitPrimitiveInlineSmallSizeWithoutREPMOVSImplRoot8(TR::Nod
...
    generateLabelInstruction(TR::InstOpCode::JMP4, node, mainEndLabel, cg);
    }
 
-static void arrayCopy8BitPrimitiveInlineSmallSizeWithoutREPMOVS(TR::Node* node, TR::Register* dstReg, TR::Register* srcReg, TR::Register* sizeReg, TR::CodeGenerator* cg)
+static void arrayCopy8BitPrimitiveInlineSmallSizeWithoutREPMOVS(TR::Node* node,
+                                                                TR::Register* dstReg,
+                                                                TR::Register* srcReg,
+                                                                TR::Register* sizeReg,
+                                                                TR::CodeGenerator* cg,
+                                                                int repMovsThresholdBytes)
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRTreeEvaluator.cpp:2188]: int32_t

File: compiler/x/codegen/OMRTreeEvaluator.cpp [hunk 7]
```diff
@@ -2266,9 +2295,11 @@ static void arrayCopyDefault(TR::Node* node, uint8_t elementSize, TR::Register*
                                                                      !disableEnhancement &&
                                                                      cg->comp()->target().cpu.supportsAVX() &&
                                                                      cg->comp()->target().is64Bit()) ? true : false;
+   int repMovsThresholdBytes = 32;
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRTreeEvaluator.cpp:2298]: int32_t

**Review by 0xdaryl — COMMENTED** [review]
Minor comments only.

**Comment by a7ehuo** [comment]
@0xdaryl All comments are addressed in Addressed in [8aa9487](https://github.com/eclipse/omr/compare/3b81ee0ea7ec4d98edb269cbea048210aa921cd6..8aa948709dc99e5837b7806172d2fbfe6e36825c). Ready for another review. Thank you!

**Comment by 0xdaryl** [comment]
Jenkins build xlinux,win,x32linux

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**